File: RealtimeDatabase/Models/Prefilter/ThenOrderByPrefilter.cs
using RealtimeDatabase.Internal;
using System;
using System.Collections.Generic;
using System.Linq;
using RealtimeDatabase.Helper;

// ReSharper disable PossibleMultipleEnumeration

namespace RealtimeDatabase.Models.Prefilter
{
    public class ThenOrderByPrefilter : IPrefilter
    {
        public string SelectFunctionString { get; set; }

        public object[] ContextData { get; set; }

        public bool Descending { get; set; }

        private Func<object, IComparable> keySelector;

        public IEnumerable<object> Execute(IEnumerable<object> array)
        {
            if (array.Any())
            {
                if (keySelector == null)
                {
                    keySelector = SelectFunctionString.CreatePredicateFunction(ContextData);
                }

                IOrderedEnumerable<object> orderedArray = (IOrderedEnumerable<object>)array;
                return Descending ? orderedArray.ThenByDescending(keySelector) : orderedArray.ThenBy(keySelector);
            }

            return array;
        }
    }
}

File: DemoClient/projects/ng-realtime-database/src/lib/models/prefilter/then-order-by-prefilter.ts
import {IPrefilter} from './iprefilter';
import {ArrayHelper} from '../../helper/array-helper';
import {OrderByPrefilter} from './order-by-prefilter';

export class ThenOrderByPrefilter<T> extends OrderByPrefilter<T> {
  prefilterType = 'ThenOrderByPrefilter';

  constructor(selectFunction: (x: T) => any, descending: boolean = false, contextData?: { [key: string]: any }) {
    super(selectFunction, descending, contextData);
  }

  public execute(values: T[]) {
    return ArrayHelper.thenOrderBy(values, this.selectFunction, this.descending);
  }
}
